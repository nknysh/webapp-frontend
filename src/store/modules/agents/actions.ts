import { ICompany, ITravelAgent } from 'services/BackendApi';
export const GET_TRAVEL_AGENTS_REQUEST = 'agentsModule/GET_TRAVEL_AGENTS_REQUEST';
export const GET_TRAVEL_AGENTS_SUCCESS = 'agentsModule/GET_TRAVEL_AGENTS_SUCCESS';
export const GET_TRAVEL_AGENTS_FAILURE = 'agentsModule/GET_TRAVEL_AGENTS_FAILURE';
export const SELECTED_TA_CHANGE = 'agentsModule/SELECTED_TA_CHANGE';
export const SHOW_TA_DROPDOWN = 'agentsModule/SHOW_TA_DROPDOWN';
export const SEARCH_TA_BY_NAME_CHANGE = 'agentsModule/SEARCH_TA_BY_NAME_CHANGE';
export const GET_COMPANIES_REQUEST = 'agentsModule/GET_COMPANIES_REQUEST';
export const GET_COMPANIES_SUCCESS = 'agentsModule/GET_COMPANIES_SUCCESS';
export const GET_COMPANIES_FAILURE = 'agentsModule/GET_COMPANIES_FAILURE';
export const SELECTED_COMPANY_CHANGE = 'agentsModule/SELECTED_COMPANY_CHANGE';
export const SHOW_COMPANY_DROPDOWN = 'agentsModule/SHOW_COMPANY_DROPDOWN';
export const SEARCH_COMPANY_BY_NAME = 'agentsModule/SEARCH_COMPANY_BY_NAME';

export type GetTravelAgentRequestAction = ReturnType<typeof getTravelAgentsRequestAction>;
export const getTravelAgentsRequestAction = (forceLoad?: boolean) => ({
  type: GET_TRAVEL_AGENTS_REQUEST as typeof GET_TRAVEL_AGENTS_REQUEST,
  forceLoad,
});

export type GetTravelAgentSuccessAction = ReturnType<typeof getTravelAgentsSuccessAction>;
export const getTravelAgentsSuccessAction = (agents: ITravelAgent[]) => ({
  type: GET_TRAVEL_AGENTS_SUCCESS as typeof GET_TRAVEL_AGENTS_SUCCESS,
  agents,
});

export type GetTravelAgentFailureAction = ReturnType<typeof getTravelAgentsFailureAction>;
export const getTravelAgentsFailureAction = (error: string) => ({
  type: GET_TRAVEL_AGENTS_FAILURE as typeof GET_TRAVEL_AGENTS_FAILURE,
  error,
});

export type SelectedTaChangeAction = ReturnType<typeof selectedTaChangeAction>;
export const selectedTaChangeAction = (value: ITravelAgent | null) => ({
  type: SELECTED_TA_CHANGE as typeof SELECTED_TA_CHANGE,
  value,
});

export type ShowTaDropdownAction = ReturnType<typeof showTaDropdownAction>;
export const showTaDropdownAction = (value: boolean) => ({
  type: SHOW_TA_DROPDOWN as typeof SHOW_TA_DROPDOWN,
  value,
});

export type SearchTaByNameChangeAction = ReturnType<typeof searchTaByNameChangeAction>;
export const searchTaByNameChangeAction = (value: string) => ({
  type: SEARCH_TA_BY_NAME_CHANGE as typeof SEARCH_TA_BY_NAME_CHANGE,
  value,
});


export type GetCompaniesRequestAction = ReturnType<typeof getCompaniesRequestAction>;
export const getCompaniesRequestAction = () => ({
  type: GET_COMPANIES_REQUEST as typeof GET_COMPANIES_REQUEST,
});

export type GetCompaniesSuccessAction = ReturnType<typeof getCompaniesSuccessAction>;
export const getCompaniesSuccessAction = (companies: ICompany[]) => ({
  type: GET_COMPANIES_SUCCESS as typeof GET_COMPANIES_SUCCESS,
  companies,
});

export type GetCompaniesFailureAction = ReturnType<typeof getCompaniesFailureAction>;
export const getCompaniesFailureAction = (error: string) => ({
  type: GET_COMPANIES_FAILURE as typeof GET_COMPANIES_FAILURE,
  error,
});

export type SelectedCompanyChangeAction = ReturnType<typeof selectedCompanyChangeAction>;
export const selectedCompanyChangeAction = (value: ICompany) => ({
  type: SELECTED_COMPANY_CHANGE as typeof SELECTED_COMPANY_CHANGE,
  value,
});

export type ShowCompanyDropdownAction = ReturnType<typeof showCompanyDropdownAction>;
export const showCompanyDropdownAction = (value: boolean) => ({
  type: SHOW_COMPANY_DROPDOWN as typeof SHOW_COMPANY_DROPDOWN,
  value,
});

export type SearchCompanyByNameAction = ReturnType<typeof searchCompanyByNameAction>;
export const searchCompanyByNameAction = (value: string) => ({
  type: SEARCH_COMPANY_BY_NAME as typeof SEARCH_COMPANY_BY_NAME,
  value,
});

export type AgentsModuleAction =
  | GetTravelAgentRequestAction
  | GetTravelAgentSuccessAction
  | GetTravelAgentFailureAction
  | SelectedTaChangeAction
  | ShowTaDropdownAction
  | SearchTaByNameChangeAction
  | GetCompaniesRequestAction
  | GetCompaniesSuccessAction
  | GetCompaniesFailureAction
  | SelectedCompanyChangeAction
  | ShowCompanyDropdownAction
  | SearchCompanyByNameAction;
