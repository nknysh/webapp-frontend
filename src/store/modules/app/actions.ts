export const SET_LATEST_APP_VERSION = 'app/SET_LATEST_APP_VERSION';

export type SetLatestAppVersion = ReturnType<typeof setLatestAppVersion>;
export const setLatestAppVersion = (latestAppVersion: string, isAppDeprecated: boolean) => ({
  type: SET_LATEST_APP_VERSION as typeof SET_LATEST_APP_VERSION,
  latestAppVersion,
  isAppDeprecated
});


export type AppAction =
 | SetLatestAppVersion
