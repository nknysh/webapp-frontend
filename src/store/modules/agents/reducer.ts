import * as Actions from './actions';
import { initialState, IAgentsModuleDomain } from './model';
import { getTaFullName } from 'store/utils';

export const agentsDomainReducer = (
  state: IAgentsModuleDomain = initialState,
  action: Actions.AgentsModuleAction
): IAgentsModuleDomain => {
  switch (action.type) {
    case Actions.GET_TRAVEL_AGENTS_REQUEST:
      return {
        ...state,
        requestPending: true,
      };

    case Actions.GET_TRAVEL_AGENTS_SUCCESS:
      return {
        ...state,
        requestPending: false,
        agents: action.agents,
      };

    case Actions.GET_TRAVEL_AGENTS_FAILURE:
      return {
        ...state,
        requestPending: true,
        agents: undefined,
        error: action.error,
      };

    case Actions.SELECTED_TA_CHANGE:
      return {
        ...state,
        selectedTa: action.value,
        taNameSearch: getTaFullName(action.value),
      };

    case Actions.SHOW_TA_DROPDOWN:
      return {
        ...state,
        showTaDropdown: action.value,
      };

    case Actions.SEARCH_TA_BY_NAME_CHANGE:
      return {
        ...state,
        taNameSearch: action.value,
      };


    case Actions.GET_COMPANIES_REQUEST:
      return {
        ...state,
        isFetchingCompanies: true,
      };

    case Actions.GET_COMPANIES_SUCCESS:
      return {
        ...state,
        isFetchingCompanies: false,
        companies: action.companies,
      };

    case Actions.GET_COMPANIES_FAILURE:
      return {
        ...state,
        isFetchingCompanies: true,
        companies: null,
        error: action.error,
      };

    case Actions.SELECTED_COMPANY_CHANGE:
      return {
        ...state,
        selectedCompany: action.value,
        companyNameSearch: action.value!.name,
        selectedTa: null
      };

    case Actions.SHOW_COMPANY_DROPDOWN:
      return {
        ...state,
        showCompanyDropdown: action.value,
      };

    case Actions.SEARCH_COMPANY_BY_NAME:
      return {
        ...state,
        companyNameSearch: action.value,
      };

    default:
      return state;
  }
};

export default agentsDomainReducer;
