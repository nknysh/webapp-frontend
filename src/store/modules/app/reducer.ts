import { initialState, AppState } from './model';
import * as Actions from './actions';
import produce from 'immer';

export const appReducer = (
  state: AppState = initialState,
  action: Actions.AppAction
) => {
  switch (action.type) {
    case Actions.SET_LATEST_APP_VERSION:
      return updateLatestAppVersionReducer(state, action);

    default:
      return state;
  }
};

export const updateLatestAppVersionReducer = (
  state: AppState,
  action: Actions.SetLatestAppVersion
): AppState => {
  return produce(state, draftState => {
    draftState.latestAppVersion = action.latestAppVersion;
    return draftState;
  });
};
