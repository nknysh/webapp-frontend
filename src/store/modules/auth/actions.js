import { prop } from 'ramda';

import { successAction, errorAction } from 'store/common/actions';

import AuthApi from 'api/auth';

export const AUTH_REQUEST = 'AUTH_REQUEST';
export const AUTH_OK = 'AUTH_OK';
export const AUTH_ERROR = 'AUTH_ERROR';
export const AUTH_RESET = 'AUTH_RESET';
export const AUTH_SET_TOKEN = 'AUTH_SET_TOKEN';
export const AUTH_SIGN_UP = 'AUTH_SIGN_UP';

// This constant is for Localstorage.
const AUTH_TOKEN = 'authToken';

const authRequest = value => ({
  type: AUTH_REQUEST,
  payload: value,
});

export const authOk = value => {
  const token = prop('token', value);
  // TODO(mark): We should use something like sagas to store this token.
  // Right now, this is called when we logIn and getUserFromToken.
  if (token) {
    localStorage.setItem(AUTH_TOKEN, token);
  }

  return {
    type: AUTH_OK,
    payload: value,
  };
};

const authError = value => {
  localStorage.removeItem(AUTH_TOKEN);

  return {
    type: AUTH_ERROR,
    payload: value,
  };
};

const authReset = value => ({
  type: AUTH_RESET,
  payload: value,
});

export const getUserFromToken = ({ token }) => dispatch => {
  dispatch(authRequest({ token }));
  return AuthApi.getUserFromToken({ token })
    .then(response => {
      const ok = { token, user: response.data };
      dispatch(authOk(ok));
    })
    .catch(error => {
      dispatch(authError(error));
      throw error;
    });
};

export const logIn = values => async dispatch => {
  dispatch(authRequest(values));
  try {
    const response = await AuthApi.logIn(values);
    const { id, user } = response.data;
    dispatch(authOk({ token: id, user }));
    return user;
  } catch (error) {
    dispatch(authError(error));
    throw error;
  }
};

export const logOut = values => dispatch => {
  localStorage.removeItem(AUTH_TOKEN);
  dispatch(authReset(values));
};

export const resetPassword = values => async dispatch => {
  dispatch(authRequest(values));

  try {
    const response = await AuthApi.resetPassword(values);
    return response;
  } catch (error) {
    dispatch(authError(error));
    throw error;
  }
};

export const setToken = token => ({
  type: AUTH_SET_TOKEN,
  payload: { token },
});

export const authSignUp = values => ({
  type: AUTH_SIGN_UP,
  payload: values,
});

export const signUp = values => dispatch => {
  dispatch(authSignUp(values));

  // This is where APi call would be handled.
  // return AuthApi.signUp(values).then(successAction).catch(errorAction);

  /**
   * @todo Return from API call the correct action and remove this
   */
  return values ? dispatch(successAction(AUTH_SIGN_UP, values)) : dispatch(errorAction(AUTH_SIGN_UP, values));
};
