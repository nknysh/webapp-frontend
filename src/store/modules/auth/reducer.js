import { createReducer, getErrorActionName, getSuccessActionName } from 'store/utils';
import { initialState, loadingReducer, successReducer, errorReducer, savingReducer } from 'store/common';

import { AUTH_TOKEN, AUTH_SET_TOKEN, AUTH_REQUEST, AUTH_SIGN_UP } from './actions';

const authSetToken = (state, { payload: { token } }) => ({
  ...state,
  token,
});

export default (state = initialState, payload) => {
  const tokenState = {
    ...state,
    token: localStorage.getItem(AUTH_TOKEN) || undefined,
  };

  return createReducer(
    {
      [AUTH_SET_TOKEN]: authSetToken,

      [AUTH_REQUEST]: loadingReducer,
      [getSuccessActionName(AUTH_REQUEST)]: successReducer,
      [getErrorActionName(AUTH_REQUEST)]: errorReducer,

      [AUTH_SIGN_UP]: savingReducer,
      [getSuccessActionName(AUTH_SIGN_UP)]: successReducer,
      [getErrorActionName(AUTH_SIGN_UP)]: errorReducer,
    },
    initialState
  )(tokenState, payload);
};
