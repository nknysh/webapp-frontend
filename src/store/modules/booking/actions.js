import { __, pick, lensProp, lensPath, prop, set, view, map, mapObjIndexed, pipe, mergeDeepRight } from 'ramda';
import { eachDay, subDays } from 'date-fns';

import client from 'api/bookings';
import { formatDate } from 'utils';

import { successAction, errorFromResponse } from 'store/common';
import { enqueueNotification } from 'store/modules/ui/actions';
import { getHotelRoom } from 'store/modules/hotels/selectors';

import { getBookingData, getBookingByHotelId, getBookingRoomDatesById } from './selectors';

const accommodationProductsLens = lensProp('accommodationProducts');
const datesLens = lensProp('dates');
const ratesLens = lensPath(['dates', 'rates']);

export const BOOKING_UPDATE = 'BOOKING_UPDATE';
export const BOOKING_SUBMIT = 'BOOKING_SUBMIT';
export const BOOKING_REMOVE = 'BOOKING_REMOVE';

export const updateBookingAction = payload => ({
  type: BOOKING_UPDATE,
  payload,
});

export const submitBookingAction = payload => ({
  type: BOOKING_SUBMIT,
  payload,
});

export const removeBooking = payload => ({
  type: BOOKING_REMOVE,
  payload,
});

export const updateBooking = (id, payload) => (dispatch, getState) => {
  const state = getState();

  dispatch(updateBookingAction(payload));
  dispatch(successAction(BOOKING_UPDATE, mergeDeepRight(getBookingData(state), { [id]: payload })));
};

export const completeBooking = (id, payload) => async (dispatch, getState) => {
  dispatch(updateBooking(id, payload));

  const state = getState();
  const booking = getBookingByHotelId(state, id);

  const finaliseAccommodationProducts = (product, productId) => {
    const hotelRoom = getHotelRoom(state, id, productId);
    const { to, from } = getBookingRoomDatesById(state, id, productId);

    const rates = prop('rates', hotelRoom);

    const days = eachDay(from, subDays(to, 1));
    const ratesForDays = pipe(
      map(formatDate),
      pick(__, rates)
    )(days);

    return pipe(
      set(datesLens, { to, from }),
      set(ratesLens, ratesForDays)
    )(product);
  };

  const finalBooking = pipe(
    set(
      accommodationProductsLens,
      mapObjIndexed(finaliseAccommodationProducts, view(accommodationProductsLens, booking))
    )
  )(booking);

  dispatch(submitBookingAction(finalBooking));

  try {
    const {
      data: { data },
    } = await client.createBooking({ data: { attributes: { id, ...finalBooking } } });

    dispatch(updateBooking(id, { reservationId: prop('uuid', data) }));
    dispatch(successAction(BOOKING_SUBMIT, data));
  } catch (e) {
    dispatch(
      enqueueNotification({ message: 'There was a problem creating your booking.', options: { variant: 'error' } })
    );
    dispatch(errorFromResponse(BOOKING_SUBMIT, e));
    throw e;
  }
};
