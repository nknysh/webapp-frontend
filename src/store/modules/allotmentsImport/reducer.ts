import { EImportEntity }  from '../importer/model';
import { importReducer } from '../importer/reducer';

export const allotmentsImport = importReducer(EImportEntity.ALLOTMENTS);
