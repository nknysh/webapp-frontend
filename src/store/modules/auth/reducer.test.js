import { signUp, authRequest } from './actions';
import reducer from './reducer';

describe('auth reducer', () => {
  it('should handle AUTH_REQUEST', () => {
    const action = authRequest({ foo: 'bar' });
    const reducedState = reducer({}, action);

    expect(reducedState).toMatchSnapshot();
  });

  it('should handle AUTH_REQUEST_SUCCESS', () => {
    const action = authRequest({ foo: 'bar' });
    const reducedState = reducer({}, action);

    expect(reducedState).toMatchSnapshot();
  });

  it('should handle AUTH_SIGN_UP', () => {
    const action = signUp({ foo: 'bar' });
    const reducedState = reducer({}, action);

    expect(reducedState).toMatchSnapshot();
  });

  it('should handle AUTH_SIGN_UP_SUCCESS', () => {
    const action = signUp({ foo: 'bar' });
    const reducedState = reducer({}, action);

    expect(reducedState).toMatchSnapshot();
  });

  it('should handle AUTH_SIGN_UP_ERROR', () => {
    const action = signUp({ foo: 'bar' });
    const reducedState = reducer({}, action);

    expect(reducedState).toMatchSnapshot();
  });
});
