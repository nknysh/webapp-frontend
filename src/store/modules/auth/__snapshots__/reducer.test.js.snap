// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages reducer should handle AUTH_SIGN_UP 1`] = `Object {}`;

exports[`pages reducer should handle AUTH_SIGN_UP_ERROR 1`] = `Object {}`;

exports[`pages reducer should handle AUTH_SIGN_UP_SUCCESS 1`] = `Object {}`;
